### PR TITLE
Fix string alloc

### PR DESCRIPTION
**Pull request by ilblackdragon**
Unsafe string allocation, for now put a slow fix to make it safe.
Need to switch reader that supports length of buffer left to optimize performance.

### CODE REVIEW

File: borsh-rs/borsh/src/de/mod.rs [hunk 1]
```diff
@@ -109,8 +109,11 @@ impl BorshDeserialize for String {
     #[inline]
     fn deserialize<R: Read>(reader: &mut R) -> Result<Self, Error> {
         let len = u32::deserialize(reader)?;
-        let mut result = vec![0; len as usize];
-        reader.read_exact(&mut result)?;
+        // TODO(16): return capacity allocation when we have the size of the buffer left from the reader.
```
Comment (nearmax) [inline borsh-rs/borsh/src/de/mod.rs]: `read_exact` is supposed to return `ErrorKind::UnexpectedEof` if there is not enough bytes to fill the buffer. Why this change is needed?

**Comment by frol** [comment]
@ilblackdragon This fix does not feel right to me. Allocating 4GB of RAM (2**32) either by `vec![0; len as usize]` or by pushing `u8` in a loop has just the same result as far as I can tell. If the allocation "fail", it will not have a chance to recover as OOM killer will come and kill the process in either case, and if it does not, `read_exact` should either fill the whole buffer or return an error.

BTW, the test case does not fail without the fix either.

**Comment by ailisp** [comment]
I think frol is right, this seems still not fixed
